src/ProtocolZoo/ProtocolZoo.jl

```diff
@@ -363,31 +295,55 @@ end
-                    put!(channel(prot.net, prot.node=>whoweswappedwith_node; permit_forward=true), msghist)
+                    if !isnothing(updategate)
+                        tag!(localslot, EntanglementHistory, newremotenode, newremoteslotid, whoweswappedwith_node, whoweswappedwith_slotidx, swappedlocal_slotidx)
+                        @debug "EntanglementTracker @$(prot.node): history=`$(history)` | message=`$msg` | Sending to $(whoweswappedwith_node).$(whoweswappedwith_slotidx)"
+                        msghist = Tag(updatetagsymbol, pastremotenode, pastremoteslotid, whoweswappedwith_slotidx, newremotenode, newremoteslotid, correction)
+                        put!(channel(prot.net, prot.node=>whoweswappedwith_node; permit_forward=true), msghist)
+                    else # We have a delete message but the qubit was swapped so add a tag and forward to swapped node
+                        @debug "EntanglementTracker @$(prot.node): history=`$(history)` | message=`$msg` | Sending to $(whoweswappedwith_node).$(whoweswappedwith_slotidx)"
+                        msghist = Tag(updatetagsymbol, pastremotenode, pastremoteslotid, whoweswappedwith_node, whoweswappedwith_slotidx)
+                        tag!(localslot, updatetagsymbol, prot.node, localslot.idx, whoweswappedwith_node, whoweswappedwith_slotidx)
+                        put!(channel(prot.net, prot.node=>whoweswappedwith_node; permit_forward=true), msghist)
+                    end
                     continue
                 end
-                error("`EntanglementTracker` on node $(prot.node) received a message $(msg) that it does not know how to handle (due to the absence of corresponding `EntanglementCounterpart` or `EntanglementHistory` tags). This is a bug in the protocol and should not happen -- please report an issue at QuantumSavory's repository.")
+
+                if !isnothing(querydelete!(localslot, EntanglementDelete, prot.node, localslot.idx, pastremotenode, pastremoteslotid)) #deletion from both sides of the swap, deletion msg when both qubits of a pair are deleted, or when EU arrives after ED at swap node with two simultaneous swaps and deletion on one side
+                    if !(isnothing(updategate)) # EU message, hence to handle a possible delete-swap-swap case, we need to update the EntanglementDelete tag
+                        tag!(localslot, EntanglementDelete, prot.node, localslot.idx, newremotenode, newremoteslotid)
+                        @debug "EntanglementTracker @$(prot.node): message=`$msg` for deleted qubit handled and EntanglementDelete tag updated"
+                    else # when the message is EntanglementDelete and the slot also has an EntanglementDelete tag(both qubits were deleted), do nothing
+                        @debug "EntanglementTracker @$(prot.node): message=`$msg` for deleted qubit handled"
+                    end
+                    continue
+                end
```

Comment:
I do not think I understand what is happening here, could you explain it in more detail (as a comment for now)?

Are you using `EntanglementDelete` both as a message sent from one node to another node to request a deletion **AND** as a historical note that a deletion has happened?

Reply:
Yes, the `EntanglementDelete` tag serves as a message and a deletion history. Looks like I forgot to mention that in the tag's docstring. Would it also be a good idea to separate all the tags used in `ProtocolZoo` module into their own file?

Reply:
I will add a comment above this line to try to describe the logic in more detail